### PR TITLE
New Torch multi-GPU programming model

### PR DESCRIPTION
**Pull request by gheinrich**
Datapoints:

MNIST+LeNet (30 epochs)
1 GPU: 56s
2 GPUs: 2m51s
(not unexpected due to communication overhead)

Upscaled CIFAR + Alexnet (10 epochs):
1 GPU: 13m11s
2 GPUs: 13m7s

Upscaled CIFAR + Googlenet (2 epochs):
1 GPU: 16m20s
2 GPUs: 11m13s

### CODE REVIEW

**Comment by lukeyeager** [comment]
Works for me. Scaling is similarly bad for MNIST. Here's the numbers on my old TitanZ (with NCCL):

MNIST+LeNet (30 epochs)
1 GPU: 1m28s
2 GPUs: 4m18s

Also, I have hundreds of lines like this in my logs:
```
BLAS : Program is Terminated. Because you tried to allocate too many memory regions.
BLAS : Program is Terminated. Because you tried to allocate too many memory regions.
BLAS : Program is Terminated. Because you tried to allocate too many memory regions.
```
The process keeps on trucking despite the error messages. Any ideas?

File: digits/standard-networks/torch/ImageNet-Training/alexnet.lua [hunk 1]
```diff
@@ -47,12 +47,18 @@ function createModel(nGPU, channels, nClasses)
...
-         cutorch.setDevice(i)
-         parallel_features:add(features:clone(), i)  -- Use the ith GPU
+      local gpus = torch.range(1, nGPU):totable()
+      local fastest, benchmark
+      local use_cudnn = cudnn ~= nil
+      if use_cudnn then
+        fastest, benchmark = cudnn.fastest, cudnn.benchmark
       end
-      cutorch.setDevice(1)  -- This is the 'primary' GPU
+      local parallel_features = nn.DataParallelTable(1, true, true):add(features,gpus):threads(function()
```
Comment (ngimel) [inline digits/standard-networks/torch/ImageNet-Training/alexnet.lua]: you may not need :threads(...) in alexnet and lenet - for small nets perhaps the overhead of threading outweighs the benefits.

**Comment by gheinrich** [comment]
Thanks for the feedback Natalia. The wall time is similar if I remove the call to `:threads(...)` on Alexnet.

**Comment by lukeyeager** [comment]
> Also, I have hundreds of lines like this in my logs:
> The process keeps on trucking despite the error messages. Any ideas?

Are you getting those error messages? That's the only thing holding me back from merging this.

**Comment by gheinrich** [comment]
If haven't seen those errors messages (neither on LeNet nor on Alexnet or GoogLeNet)

**Comment by lukeyeager** [comment]
I'm not seeing the warnings now. I'll go ahead and merge this - the warnings weren't causing real problems anyway.

**Pull request merged**